API functions for the new generic VerificationAttempt model in the verify_student app

ilee2u opened the pull request:
## Description

This PR adds api functions which can be called by IDV plugins to create and update instances of the recently added generic VerificationAttempt model.

Context for this decision can be found in this excerpt from our internal spec doc:

```
We will introduce one new generic model - the VerificationAttempt model. The
purpose of this model is to store data about verification attempts on the
platform. 

The VerificationAttempt model will store information about each IDV attempt. It
will have the majority, if not all, the fields on the
[IDVerificationAttempt](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/verify_student/models.py#L94) abstract base model. The biggest change is that the set of
available statuses will be created, pending, approved, and denied. 

Eventually, the vision is for all forms of IDV to be installed into the platform
as plugins. The plugins would write to the VerificationAttempt model so that the
platform has a record of the IDV information it needs. The rationale is that the
platform only cares whether the learner is ID verified and not how they are ID
verified.

This would replace the majority of filter hooks proposed in the original spec.
This is because doing database queries via filter hooks can cause unpredictable
performance problems. This has been a pain point for Open edX. Write now, read
later is a better pattern to follow.

Note that the IDVerificationURLRequested(url) filter hook is still a part of the
design to replace the IDV URL on the platform. Although we do not need to use it
to integrate Persona, it is necessary to enable additional implementations of
IDV via these changes.
```

Design decisions and their rationales should be documented in the repo (docstring / ADR), per
[OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:

- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
"Developer", and "Operator".
  - Learners will primarily be impacted due to our primary IDV vendor changing.
  - Course Authors will also be somewhat impacted, at the very least because they will need to be notified of the changed IDV process (which we've already made efforts to do).
  - Developers and Operators will also be somewhat impacted, as they'll also need to be notified of this change and how this impacts support as well as future development on IDV and IDV-related systems.


## Supporting information

Info from internal ticket:

introduce a Python API for the generic VerificationAttempt model in the [verify_student](https://github.com/openedx/edx-platform/tree/master/lms/djangoapps/verify_student) application. The purpose of this API is to allow plugins that implement different types of IDV to import these methods and effect changes in the [verify_student](https://github.com/openedx/edx-platform/tree/master/lms/djangoapps/verify_student) VerificationAttempt model.

Below is an idea for the function signatures.

```
id = create_verification_attempt(user, name, status, expiration_date=None)
update_verification_attempt(id, status)
```

Note that an id must be returned from create_verification_attempt so that the plugin can specify the id on any following update_verification calls.

## Deadline

None

## Other information

Other changes will be made in `openedx-events` and `edx-name-affirmation` to implement the downstream effects of calling these new API functions.

michaelroytman committed:
feat: add VerificationAttempt model to verify_student application

This commits adds a VerificationAttempt model to store implementation and provider agnostic information about identity verification attempts in the platform.

ilee2u committed:
feat: add api for VerificationAttempt model

varshamenon4 commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+    Create a verification attempt.
+
+    Args:
+        user (User): the user (usually a learner) performing the verfication attempt
```
Nit:
```suggestion
        user (User): the user (usually a learner) performing the verification attempt
```

varshamenon4 commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+
+    Arguments:
+        * id (str): the verification attempt id
+        * status (str): the new status
```
Does there need to be validation that the status is correct here?

ilee2u replied:
@MichaelRoytman is there any reason you could think of that we'd need to check that the status in `verify_student`'s generic `VerificationAttempt` model went through a valid status transition? My guess is that, in the long term, all instances of a VerificationAttempt (whether they use Persona or not) will eventually have their own checks for valid status transitions, since each IDV process uses different statuses.

varshamenon4 replied:
That makes sense to me Isaac! Wasn't sure if this needed to happen here, especially if it's already happening elsewhere. :)

michaelroytman replied:
I think that we should check that `status` is a valid status in both the `create` and `update` methods (i.e. it is one of `created`, `pending`, `approved`, or `denied`).

In terms of status transitions, I agree that it would be tough to assert a particular set of valid state transitions here, because each form of IDV will be different.

varshamenon4 reviewed (commented):
Looks good so far! Just a couple of nits and questions.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+
+    Args:
+        user (User): the user (usually a learner) performing the verfication attempt
+        name (string): the name of the user
```
This should be the name being ID verified. It's not necessarily the current user's name.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+        expiration_datetime (datetime, optional): When the verification attempt expires. Defaults to None.
+
+    Returns:
+        VerificationAttempt (VerificationAttempt): The created VerificationAttempt instance
```
You're returning the `id` field, so could you update this comment?

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+
+def create_verification_attempt(user, name, status, expiration_datetime=None):
+    """
+    Create a verification attempt.
```
I think we should document that this method is intended to be used by IDV implementation plugins to create `VerificationAttempts` - same for `update_attempt_status`.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+    Update the VerificationAttempt status.
+
+    Arguments:
+        * id (str): the verification attempt id
```
```suggestion
        * id (str): the verification attempt id of the attempt to update
```

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+    Returns:
+        * None
+    """
+    attempt = VerificationAttempt.objects.get(id=attempt_id)
```
This may throw a `VerificationAttempt.NotFound` exception if the `get` call fails. Could you handle that case?

michaelroytman commented on lms/djangoapps/verify_student/management/commands/approve_id_verifications.py:
```diff
@@ -8,7 +8,7 @@
 from pprint import pformat
 
-from django.contrib.auth.models import User  # lint-amnesty, pylint: disable=imported-auth-user
+from django.contrib.auth.models import User  # lint-amnesty, pylint: disable=imported-auth-user, unused-import
```
If this import isn't used, could we remove it instead of ignoring the lint error?

michaelroytman replied:
Bump.

michaelroytman commented on lms/djangoapps/verify_student/tests/test_api.py:
```diff
@@ -41,3 +48,77 @@ def test_send_approval(self, use_ace):
+@ddt.ddt
+class CreateVerificationAttempt(TestCase):
+    """
+    Test cases for the send_approval_email API method.
```
Please update this docstring to document which method you're testing.

michaelroytman commented on lms/djangoapps/verify_student/tests/test_api.py:
```diff
@@ -41,3 +48,77 @@ def test_send_approval(self, use_ace):
+        )
+        self.attempt.save()
+
+    def test_create_verification_attempt(self):
```
Let's test when `expiration_datetime` is not supplied, as well.

michaelroytman commented on lms/djangoapps/verify_student/tests/test_api.py:
```diff
@@ -41,3 +48,77 @@ def test_send_approval(self, use_ace):
+        VerificationAttemptStatus.approved,
+        VerificationAttemptStatus.denied,
+    )
+    def test_update_verification_status(self, to_status):
```
When you handle the case where the `get` call fails and a `VerificationAttempt.NotFound` exception is raised, let's also test that case.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+
+def update_verification_status(attempt_id, status):
+    """
+    Update the VerificationAttempt status.
```
```suggestion
    Update the status of a verification attempt.
```

Nit: to be parallel with `create_verification_attempt`.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +34,42 @@ def send_approval_email(attempt):
+    return verification_attempt.id
+
+
+def update_verification_status(attempt_id, status):
```
```suggestion
def update_verification_attempt_status(attempt_id, status):
```

Nit: to be parallel with `create_verification_attempt`.

ilee2u committed:
fix: error handling for update

- added tests accordingly
- also took care of some nits

ilee2u committed:
chore: lint

ilee2u committed:
chore: lint for equals spaces

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,63 @@ def send_approval_email(attempt):
+    Returns:
+        * None
+    """
+    status_list = [attr for attr in dir(VerificationAttemptStatus) if not attr.startswith('__')]
```
My concern here is that this is brittle. If any other attributes are added to the class, they'll be treated as statuses - even methods!

I see three alternatives.

1. We could enumerate the list of valid statuses here.
2. We could add an attribute like `all_statuses` to the `VerificationAttemptStatus` class.
3. We could add an `is_valid_status` to the `VerificationAttemptStatus` class that `update_verification_attempt_status` class calls.

I'm partial to #3, because then the status class is responsible for defining what a valid status is, not the API function.

What do you think?

michaelroytman replied:
In retrospect, trying to save a non-valid status (i.e. something that isn't in the `choices` list on the model) will raise a `ValidationError`, so Django handles this for us out of the box. You could consider keeping the default Django behavior instead of explicitly checking the status. My mistake.

ilee2u replied:
Added a comment below with how I decided to handle this: https://github.com/openedx/edx-platform/pull/35338#discussion_r1735144386

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,63 @@ def send_approval_email(attempt):
+    if status not in status_list:
+        log.error(
+            f'update_verification_attempt_status called with invalid status. '
+            f'Status must be one of: {status_list}',
```
I'd recommend also adding the status it was called with. This will be helpful to know so we can track down bugs more easily.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,63 @@ def send_approval_email(attempt):
+            f'VerificationAttempt with id {attempt_id} was not found '
+            f'when updating the attempt to status={status}',
+        )
+        raise VerificationAttempt.DoesNotExist  # pylint: disable=raise-missing-from
```
If you're simply re-raising the same exception you caught, then it's best practice to just use `raise`. This will do what you're doing here and will avoid the pylint error, so you could remove the disable line.

michaelroytman commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -1213,4 +1213,5 @@ class VerificationAttempt(TimeStampedModel):
     expiration_datetime = models.DateTimeField(
         null=True,
         blank=True,
+        db_index=True,
```
I removed this index in my pull request. Could you remove this line, please?

michaelroytman reviewed (commented):
Looking good! I left a few more comments.

Could you also update the PR description and answer the prompts so that others can better review your changes? Thanks!

ilee2u committed:
feat: using generic update function instead

- can now update name, status, and exp. date on generic attempts
- changed tests accordingly
- a few nits

ilee2u committed:
chore: fix docstring args

ilee2u commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,76 @@ def send_approval_email(attempt):
+    try:
+        attempt.clean_fields()
+        attempt.save()
+    except ValidationError:
```
I had to add this call to `attempt.clean_fields()` in order for this ValidationError to work as @MichaelRoytman mentioned in their previous comment. Let me know if there's something else I should do that's best practice.

zacharis278 replied:
this reads like it would take _any_ validation error and report back that the status is invalid. If we're only checking for bad status we should do that a bit more explicitly. I'd fear this would create confusing behavior where another field fails to validate but our system incorrectly tells us the status is bad.

michaelroytman replied:
We should probably just call `full_clean()` here to validate the changes fully if we take this approach.

That said, maybe it would be easier to do what you had before and validate the status values in Python? I didn't realize that calling `attempt.save()` wouldn't perform that validation.

I haven't seen the use of Django forms a lot at edX, but using a form would perform the validation automatically, because calling `save()` on a form will call `full_clean()`.

But I agree with Zach. The validation error will look something like this, so you could key off that.

```
django.core.exceptions.ValidationError: {'status': ["Value 'invalid' is not a valid choice."]}
```

ilee2u replied:
I put back the code to only validate which I had earlier

michaelroytman commented on lms/djangoapps/verify_student/tests/test_api.py:
```diff
@@ -41,3 +49,140 @@ def test_send_approval(self, use_ace):
+            status=VerificationAttemptStatus.approved,
+        )
+
+    # These are statuses used in edx-name-affirmation's VerifiedName model and persona-integration's unique
```
I think we should remove this comment, especially about `persona-integration`, which we shouldn't mention in the `edx-platform`, since it's a private plugin.

I don't think you need to explain where these statuses come from - testing invalid statuses is a normal thing to do, so it's justified as is!

ilee2u committed:
fix: corrected status validation

- reverted to old status validation method
- fixed tests accordingly

ormsbee commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,76 @@ def send_approval_email(attempt):
+            )
+            raise VerificationAttemptInvalidStatus
+
+    if expiration_datetime is not None:
```
Is it the case that if `VerificationAttempt.expiration_datetime` is `None`, that means "this verification never expires"? If that's the case, then if a `expiration_datetime` was set to be sometime next year, and I wanted to update it to never expire, is that possible?

ilee2u replied:
If the value of the `expiration_datetime` parameter sent to this function is `None`, then the `expiration_datetime` of the IDV attempt will _not be updated_. This is to maintain that all IDV attempts will expire/retire at some point (which at least for IDV attempts via Persona, should be 14 days).

michaelroytman replied:
Historically, we have set the expiration date from the time that an IDV attempt is marked `approved`. For this reason, I suggested that we make `expiration_datetime` optional because it may not make sense for a non-approved IDV attempt to have an expiration date. For example, if you were to update an instance of `VerificationAttempt` from the `status` `created` to `pending`, there should be no expiration date.

However, this API should remain flexible, because that depends on a particular implementation of IDV. I'd agree that we should be able to update the `expiration_datetime` for any status.

Isaac, to Dave's point, I think that we shouldn't prevent the user of this function from setting `expiration_datetime` to `None`, necessarily. We should just not require that they supply a value for the parameter. What you think?

ilee2u replied:
That makes sense to me, thank you for clarifying that! I'll push this change soon.

ormsbee commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,76 @@ def send_approval_email(attempt):
+    if status is not None:
+        attempt.status = status
+
+        status_list = [attr for attr in dir(VerificationAttemptStatus) if not attr.startswith('__')]
```
Nit (optional): FWIW, Python 3.11 has [StrEnum](https://docs.python.org/3.11/library/enum.html#enum.StrEnum) now, which might be a more natural way to express this.

ilee2u replied:
I attempted to use StrEnum for this, however that involved needing to change the `VerificationAttemptStatus` class itself, which in turn would require me to refactor quite a bit in this application (especially within tests) so this unfortunately doesn't seem worth it.

zacharis278 replied:
a pattern I've used for this in the past is to give my constant 'enum' class a list property with valid values. Eg https://github.com/edx/edx-exams/blob/1b76cf33d8b0fbca6e15742cad8ab49f02960f1d/edx_exams/apps/core/statuses.py#L54

That, or can't we just use `VerificationAttempt.status_choices`?

michaelroytman replied:
At a minimum, I'd much prefer that we be explicit about the statuses in this list (i.e. `status_list = [VerificationAttemptStatus.created, ...]`) over implicit. This is brittle and will stick anything that isn't a dunder method into this list, including any future attributes or methods.

But I do not think that converting that class to use the Enum module would be that hard. Where is this class used in tests? I currently only see it in use in the `VerificationAttempt` model.

```
from enum import StrEnum, auto

class VerificationAttemptStatus(StrEnum):
    CREATED = auto()
    PENDING = auto()
    APPROVED = auto()
    DENIED = auto()
```

Here, you could do the following...
```
status_list = list(VerificationAttemptStatus)
```

You'd have to update `STATUS_CHOICES` in the model as well. I'm not sure whether that would require a migration.

My only concern is that we sometimes like to add methods to these status classes, like `is_terminal_status`, for example. But we don't seem to have a need for that right now, and this does really work like an Enum, so it's probably fine to make this change.

Please let me know if I'm missing something about the difficulty of this change.

ilee2u replied:
Oh I see what happened, I misread the docs and had `VerificationAttemptStatus` inherit from `enum.Flag` instead of `enum.StrEnum` 🤦. I implemented what you suggested, and I didn't have to make any migrations, and everything seems to work just fine. Thanks so much!

ormsbee commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -33,3 +40,76 @@ def send_approval_email(attempt):
         send_verification_approved_email(context=email_context)
+
+
+def create_verification_attempt(user, name, status, expiration_datetime=None):
```
Nit (optional): A number of APIs use type annotations these days, which can be helpful for documentation and correctness checking.

ilee2u replied:
I've added annotations to these functions!

ormsbee reviewed (commented):
Minor comments/nits for your consideration. Nothing blocking. Thank you for your patience with my late review.

michaelroytman reviewed (commented):
I left some comments. I'll clear my change request so that it's not blocking. However, even if you choose not to use the `StrEnum` class, I think that we should address the way `status_list` is constructed to be more explicit before merging.

ilee2u committed:
fix: datetime, status, and annotation fixes

- expiration_datetime can be updated to None
- VerificationAttemptStatus is now StrEnum
- Added type annotations for api functions

ormsbee reviewed (approved):
Thank you!
